File: backend/back_api/src/back_api/schemas/compare.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field

from back_api.schemas.backtest import MarkowitzParams


class StrategyConfig(BaseModel):
    label: str = ""
    strategy: Literal["momentum", "mean_reversion", "orchestrator", "markowitz"] = "orchestrator"
    markowitz: MarkowitzParams = Field(default_factory=MarkowitzParams)


class CompareRequest(BaseModel):
    data_dir: str = ""
    assets: dict[str, str] = Field(
        default={"BTC": "BTCUSDT", "ETH": "ETHUSDT"},
    )
    freq: str = Field(default="1D")
    min_train_bars: int = Field(default=252, ge=10)
    rebalance_every: int = Field(default=21, ge=1)
    fee_rate: float = Field(default=0.001, ge=0.0)
    slippage: float = Field(default=0.0005, ge=0.0)
    strategies: list[StrategyConfig] = Field(min_length=1, max_length=10)
    include_frontier: bool = True
    frontier_points: int = Field(default=20, ge=5, le=100)
    frontier_cov_model: Literal["rolling", "diagonal", "bekk"] = "diagonal"
    frontier_er_model: Literal["rolling_mean", "signal_tsmom"] = "signal_tsmom"
    frontier_long_only: bool = False


class StrategyResultEntry(BaseModel):
    label: str
    strategy: str
    equity_curve: list[tuple[str, float]]
    metrics: dict[str, float]
    trade_count: int
    signals: list[tuple[str, dict[str, float]]]
    ann_return: float
    ann_vol: float


class FrontierPoint(BaseModel):
    gamma: float
    ann_return: float
    ann_vol: float
    sharpe: float
    weights: dict[str, float]


class CompareResult(BaseModel):
    strategies: list[StrategyResultEntry]
    benchmarks: dict[str, list[tuple[str, float]]]
    frontier: list[FrontierPoint] | None = None
    frontier_special: dict[str, FrontierPoint] | None = None


class CompareJobResponse(BaseModel):
    job_id: str
    status: str
    result: CompareResult | None = None
    error: str | None = None

File: frontend/src/types/compare.ts
import type { StrategyName, MarkowitzParams } from './backtest'

export interface StrategyConfig {
  label: string
  strategy: StrategyName
  markowitz?: MarkowitzParams
}

export interface CompareRequest {
  freq?: string
  min_train_bars: number
  rebalance_every: number
  fee_rate: number
  slippage: number
  strategies: StrategyConfig[]
  include_frontier: boolean
  frontier_points: number
  frontier_cov_model: 'rolling' | 'diagonal' | 'bekk'
  frontier_er_model: 'rolling_mean' | 'signal_tsmom'
  frontier_long_only: boolean
}

export interface StrategyResultEntry {
  label: string
  strategy: string
  equity_curve: [string, number][]
  metrics: Record<string, number>
  trade_count: number
  signals: [string, Record<string, number>][]
  ann_return: number
  ann_vol: number
}

export interface FrontierPoint {
  gamma: number
  ann_return: number
  ann_vol: number
  sharpe: number
  weights: Record<string, number>
}

export interface CompareResult {
  strategies: StrategyResultEntry[]
  benchmarks: Record<string, [string, number][]>
  frontier: FrontierPoint[] | null
  frontier_special: Record<string, FrontierPoint> | null
}

export type CompareJobStatus = 'pending' | 'running' | 'done' | 'failed'

export interface CompareJobResponse {
  job_id: string
  status: CompareJobStatus
  result?: CompareResult
  error?: string
}
